src/bin/common/parse_command.ml

```diff
@@ -895,32 +853,171 @@ let parse_output_opt =
-    Arg.(value & opt interpretation_conv INone &
-         info ["interpretation"] ~docv ~docs ~doc) in
+  (* Use the --interpretation and --produce-models (which is equivalent to
+     --interpretation last) to determine the interpretation value. *)
+  let interpretation, dump_models =
+    let interpretation =
+      let doc = Format.sprintf
+          "Best effort support for counter-example generation. \
+           $(docv) must be %s. %s shows the first computed interpretation. \
+           %s compute an interpretation before every decision, \
+           and %s only before returning unknown. \
+           Note that $(b, --max-split) limitation will \
+           be ignored in model generation phase."
+          (Arg.doc_alts
+             ["none"; "first"; "every"; "last"])
+          (Arg.doc_quote "first") (Arg.doc_quote "every")
+          (Arg.doc_quote "last") in
+      let docv = "VAL" in
+      let interpretation =
+        Arg.enum
+          [ "none", INone
+          ; "first", IFirst
+          ; "every", IEvery
+          ; "last", ILast
+          ]
+      in
+      Arg.(value & opt interpretation INone &
+           info ["interpretation"] ~docv ~docs:s_models ~doc)
+    in
+
+    let produce_models =
+      let doc =
+        "Enable model generation (equivalent to --interpretation last)."
+      in
+      Arg.(value & flag & info ["produce-models"] ~doc ~docs:s_models)
+    in
+
+    let dump_models =
+      let doc =
+        "Display a model each time the result is unknown (implies \
+         --interpretation last)"
+      in
+      Arg.(value & flag & info ["dump-models"] ~doc ~docs:s_models)
+    in
+
+    let mk_interpretation interpretation produce_models dump_models =
+      match interpretation with
+      | INone when produce_models || dump_models -> ILast
+      | interpretation -> interpretation
+    in
+    Term.(
+      const mk_interpretation $ interpretation $
+      produce_models $ dump_models
+    ),
+    dump_models
+  in
+
+  (* Use the --sat-solver to determine the sat solver.
+
+     If an interpretation is provided, the solver is forced to be Tableaux,
+     because generation of models requires OptimAE for the other solvers.
+
+     See https://github.com/OCamlPro/alt-ergo/pull/553 *)
+  let sat_solver =
+    let sat_solver_arg =
+      let sat_solver : _ Arg.conv =
+        let parse = function
+          | "CDCL" | "satML" ->
+            Ok Util.CDCL
+          | "CDCL-Tableaux" | "satML-Tableaux"
+          | "CDCL-tableaux" | "satML-tableaux" ->
+            Ok Util.CDCL_Tableaux
+          | "tableaux" | "Tableaux"
+          | "tableaux-like" | "Tableaux-like" ->
+            Ok Util.Tableaux
+          | "tableaux-cdcl" | "Tableaux-CDCL"
+          | "tableaux-CDCL" | "Tableaux-cdcl" ->
+            Ok Util.Tableaux_CDCL
+          | sat_solver ->
+            Error ("Args parsing error: unkown SAT solver " ^ sat_solver)
+
+        in
+        Arg.(conv' (parse, Util.pp_sat_solver))
```

Comment:
You can use the predefined converter `enum`. Uhm but we cannot manage alternative spellings with the predefined converter. In my opinion we should allow the capitalized versions.

Reply:
Yeah, I didn't use `enum` because I wanted to preserve the old lax behavior. Did you mean "we should *not* allow the capitalized version"?

One intermediate solution would be to keep the current code but choose a single name for each solver and emit a deprecation warning if other cases are used, then switch to `enum` and remove the deprecated cases in the next release.

Reply:
OK, I'll merge this and make a separate PR with the deprecation!